File: backend/crates/pytake-core/src/auth/permissions.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

/// User roles in the system
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    Admin,
    Supervisor,
    Agent,
    Viewer,
}

impl Role {
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "admin" => Some(Role::Admin),
            "supervisor" => Some(Role::Supervisor),
            "agent" => Some(Role::Agent),
            "viewer" => Some(Role::Viewer),
            _ => None,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Role::Admin => "admin",
            Role::Supervisor => "supervisor",
            Role::Agent => "agent",
            Role::Viewer => "viewer",
        }
    }

    /// Get all permissions for this role
    pub fn permissions(&self) -> HashSet<Permission> {
        match self {
            Role::Admin => Permission::all(),
            Role::Supervisor => Permission::supervisor_permissions(),
            Role::Agent => Permission::agent_permissions(),
            Role::Viewer => Permission::viewer_permissions(),
        }
    }

    /// Check if role has a specific permission
    pub fn has_permission(&self, permission: Permission) -> bool {
        self.permissions().contains(&permission)
    }
}

/// System permissions
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Permission {
    // User management
    ViewUsers,
    CreateUsers,
    UpdateUsers,
    DeleteUsers,
    
    // Conversation management
    ViewConversations,
    CreateConversations,
    UpdateConversations,
    DeleteConversations,
    AssignConversations,
    
    // Message management
    ViewMessages,
    SendMessages,
    DeleteMessages,
    
    // Contact management
    ViewContacts,
    CreateContacts,
    UpdateContacts,
    DeleteContacts,
    
    // Template management
    ViewTemplates,
    CreateTemplates,
    UpdateTemplates,
    DeleteTemplates,
    
    // Media management
    ViewMedia,
    UploadMedia,
    DeleteMedia,
    
    // Dashboard and reports
    ViewDashboard,
    ViewReports,
    ExportData,
    
    // Organization management
    ViewOrganization,
    UpdateOrganization,
    ManageIntegrations,
    
    // System administration
    ViewSystemSettings,
    UpdateSystemSettings,
    ViewAuditLogs,
    ManageWebhooks,
}

impl Permission {
    /// Get all permissions (for admin)
    pub fn all() -> HashSet<Permission> {
        use Permission::*;
        vec![
            ViewUsers, CreateUsers, UpdateUsers, DeleteUsers,
            ViewConversations, CreateConversations, UpdateConversations, DeleteConversations, AssignConversations,
            ViewMessages, SendMessages, DeleteMessages,
            ViewContacts, CreateContacts, UpdateContacts, DeleteContacts,
            ViewTemplates, CreateTemplates, UpdateTemplates, DeleteTemplates,
            ViewMedia, UploadMedia, DeleteMedia,
            ViewDashboard, ViewReports, ExportData,
            ViewOrganization, UpdateOrganization, ManageIntegrations,
            ViewSystemSettings, UpdateSystemSettings, ViewAuditLogs, ManageWebhooks,
        ].into_iter().collect()
    }

    /// Get supervisor permissions
    pub fn supervisor_permissions() -> HashSet<Permission> {
        use Permission::*;
        vec![
            ViewUsers, CreateUsers, UpdateUsers,
            ViewConversations, CreateConversations, UpdateConversations, AssignConversations,
            ViewMessages, SendMessages, DeleteMessages,
            ViewContacts, CreateContacts, UpdateContacts, DeleteContacts,
            ViewTemplates, CreateTemplates, UpdateTemplates, DeleteTemplates,
            ViewMedia, UploadMedia, DeleteMedia,
            ViewDashboard, ViewReports, ExportData,
            ViewOrganization,
            ViewAuditLogs,
        ].into_iter().collect()
    }

    /// Get agent permissions
    pub fn agent_permissions() -> HashSet<Permission> {
        use Permission::*;
        vec![
            ViewConversations, UpdateConversations,
            ViewMessages, SendMessages,
            ViewContacts, CreateContacts, UpdateContacts,
            ViewTemplates, CreateTemplates,
            ViewMedia, UploadMedia,
            ViewDashboard,
        ].into_iter().collect()
    }

    /// Get viewer permissions
    pub fn viewer_permissions() -> HashSet<Permission> {
        use Permission::*;
        vec![
            ViewConversations,
            ViewMessages,
            ViewContacts,
            ViewTemplates,
            ViewMedia,
            ViewDashboard,
        ].into_iter().collect()
    }

    /// Get permission from string
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "view_users" => Some(Permission::ViewUsers),
            "create_users" => Some(Permission::CreateUsers),
            "update_users" => Some(Permission::UpdateUsers),
            "delete_users" => Some(Permission::DeleteUsers),
            "view_conversations" => Some(Permission::ViewConversations),
            "create_conversations" => Some(Permission::CreateConversations),
            "update_conversations" => Some(Permission::UpdateConversations),
            "delete_conversations" => Some(Permission::DeleteConversations),
            "assign_conversations" => Some(Permission::AssignConversations),
            "view_messages" => Some(Permission::ViewMessages),
            "send_messages" => Some(Permission::SendMessages),
            "delete_messages" => Some(Permission::DeleteMessages),
            "view_contacts" => Some(Permission::ViewContacts),
            "create_contacts" => Some(Permission::CreateContacts),
            "update_contacts" => Some(Permission::UpdateContacts),
            "delete_contacts" => Some(Permission::DeleteContacts),
            "view_templates" => Some(Permission::ViewTemplates),
            "create_templates" => Some(Permission::CreateTemplates),
            "update_templates" => Some(Permission::UpdateTemplates),
            "delete_templates" => Some(Permission::DeleteTemplates),
            "view_media" => Some(Permission::ViewMedia),
            "upload_media" => Some(Permission::UploadMedia),
            "delete_media" => Some(Permission::DeleteMedia),
            "view_dashboard" => Some(Permission::ViewDashboard),
            "view_reports" => Some(Permission::ViewReports),
            "export_data" => Some(Permission::ExportData),
            "view_organization" => Some(Permission::ViewOrganization),
            "update_organization" => Some(Permission::UpdateOrganization),
            "manage_integrations" => Some(Permission::ManageIntegrations),
            "view_system_settings" => Some(Permission::ViewSystemSettings),
            "update_system_settings" => Some(Permission::UpdateSystemSettings),
            "view_audit_logs" => Some(Permission::ViewAuditLogs),
            "manage_webhooks" => Some(Permission::ManageWebhooks),
            _ => None,
        }
    }

    /// Convert permission to string
    pub fn as_str(&self) -> &'static str {
        match self {
            Permission::ViewUsers => "view_users",
            Permission::CreateUsers => "create_users",
            Permission::UpdateUsers => "update_users",
            Permission::DeleteUsers => "delete_users",
            Permission::ViewConversations => "view_conversations",
            Permission::CreateConversations => "create_conversations",
            Permission::UpdateConversations => "update_conversations",
            Permission::DeleteConversations => "delete_conversations",
            Permission::AssignConversations => "assign_conversations",
            Permission::ViewMessages => "view_messages",
            Permission::SendMessages => "send_messages",
            Permission::DeleteMessages => "delete_messages",
            Permission::ViewContacts => "view_contacts",
            Permission::CreateContacts => "create_contacts",
            Permission::UpdateContacts => "update_contacts",
            Permission::DeleteContacts => "delete_contacts",
            Permission::ViewTemplates => "view_templates",
            Permission::CreateTemplates => "create_templates",
            Permission::UpdateTemplates => "update_templates",
            Permission::DeleteTemplates => "delete_templates",
            Permission::ViewMedia => "view_media",
            Permission::UploadMedia => "upload_media",
            Permission::DeleteMedia => "delete_media",
            Permission::ViewDashboard => "view_dashboard",
            Permission::ViewReports => "view_reports",
            Permission::ExportData => "export_data",
            Permission::ViewOrganization => "view_organization",
            Permission::UpdateOrganization => "update_organization",
            Permission::ManageIntegrations => "manage_integrations",
            Permission::ViewSystemSettings => "view_system_settings",
            Permission::UpdateSystemSettings => "update_system_settings",
            Permission::ViewAuditLogs => "view_audit_logs",
            Permission::ManageWebhooks => "manage_webhooks",
        }
    }
}

/// Permission check helper
pub struct PermissionChecker {
    user_role: Role,
    user_permissions: HashSet<Permission>,
}

impl PermissionChecker {
    pub fn new(role: Role) -> Self {
        Self {
            user_role: role,
            user_permissions: role.permissions(),
        }
    }

    pub fn with_custom_permissions(role: Role, permissions: HashSet<Permission>) -> Self {
        Self {
            user_role: role,
            user_permissions: permissions,
        }
    }

    pub fn has_permission(&self, permission: Permission) -> bool {
        self.user_permissions.contains(&permission)
    }

    pub fn has_any_permission(&self, permissions: &[Permission]) -> bool {
        permissions.iter().any(|p| self.has_permission(*p))
    }

    pub fn has_all_permissions(&self, permissions: &[Permission]) -> bool {
        permissions.iter().all(|p| self.has_permission(*p))
    }

    pub fn get_role(&self) -> Role {
        self.user_role
    }

    pub fn is_admin(&self) -> bool {
        self.user_role == Role::Admin
    }

    pub fn is_supervisor_or_above(&self) -> bool {
        matches!(self.user_role, Role::Admin | Role::Supervisor)
    }

    pub fn is_agent_or_above(&self) -> bool {
        matches!(self.user_role, Role::Admin | Role::Supervisor | Role::Agent)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_role_permissions() {
        let admin = Role::Admin;
        assert!(admin.has_permission(Permission::DeleteUsers));
        assert!(admin.has_permission(Permission::ViewDashboard));

        let agent = Role::Agent;
        assert!(agent.has_permission(Permission::SendMessages));
        assert!(!agent.has_permission(Permission::DeleteUsers));

        let viewer = Role::Viewer;
        assert!(viewer.has_permission(Permission::ViewMessages));
        assert!(!viewer.has_permission(Permission::SendMessages));
    }

    #[test]
    fn test_permission_checker() {
        let checker = PermissionChecker::new(Role::Supervisor);
        
        assert!(checker.has_permission(Permission::CreateUsers));
        assert!(!checker.has_permission(Permission::UpdateSystemSettings));
        
        assert!(checker.has_any_permission(&[
            Permission::UpdateSystemSettings,
            Permission::ViewUsers,
        ]));
        
        assert!(!checker.has_all_permissions(&[
            Permission::ViewUsers,
            Permission::UpdateSystemSettings,
        ]));
    }

    #[test]
    fn test_role_hierarchy_checks() {
        let admin_checker = PermissionChecker::new(Role::Admin);
        assert!(admin_checker.is_admin());
        assert!(admin_checker.is_supervisor_or_above());
        assert!(admin_checker.is_agent_or_above());

        let agent_checker = PermissionChecker::new(Role::Agent);
        assert!(!agent_checker.is_admin());
        assert!(!agent_checker.is_supervisor_or_above());
        assert!(agent_checker.is_agent_or_above());
    }
}

File: frontend/src/utils/permissions.ts
import type { Role, Permission } from '@/types/auth'

// Define permissions for each role
export const ROLE_PERMISSIONS: Record<Role, Permission[]> = {
  Admin: [
    // User Management
    'ViewUsers', 'CreateUsers', 'UpdateUsers', 'DeleteUsers',
    // Conversation Management
    'ViewConversations', 'CreateConversations', 'UpdateConversations', 'DeleteConversations', 'AssignConversations',
    // Message Management
    'ViewMessages', 'SendMessages', 'DeleteMessages',
    // Contact Management
    'ViewContacts', 'CreateContacts', 'UpdateContacts', 'DeleteContacts',
    // Template Management
    'ViewTemplates', 'CreateTemplates', 'UpdateTemplates', 'DeleteTemplates',
    // Media Management
    'ViewMedia', 'UploadMedia', 'DeleteMedia',
    // System & Analytics
    'ViewDashboard', 'ViewReports', 'ExportData', 'ViewOrganization', 'UpdateOrganization',
    'ManageIntegrations', 'ViewSystemSettings', 'UpdateSystemSettings', 'ViewAuditLogs', 'ManageWebhooks'
  ],
  
  Supervisor: [
    // User Management (limited)
    'ViewUsers', 'CreateUsers', 'UpdateUsers',
    // Conversation Management
    'ViewConversations', 'CreateConversations', 'UpdateConversations', 'DeleteConversations', 'AssignConversations',
    // Message Management
    'ViewMessages', 'SendMessages', 'DeleteMessages',
    // Contact Management
    'ViewContacts', 'CreateContacts', 'UpdateContacts', 'DeleteContacts',
    // Template Management
    'ViewTemplates', 'CreateTemplates', 'UpdateTemplates', 'DeleteTemplates',
    // Media Management
    'ViewMedia', 'UploadMedia', 'DeleteMedia',
    // System & Analytics
    'ViewDashboard', 'ViewReports', 'ExportData', 'ViewAuditLogs'
  ],
  
  Agent: [
    // Conversation Management (limited)
    'ViewConversations', 'UpdateConversations',
    // Message Management
    'ViewMessages', 'SendMessages',
    // Contact Management
    'ViewContacts', 'CreateContacts', 'UpdateContacts',
    // Template Management (basic)
    'ViewTemplates', 'CreateTemplates',
    // Media Management (basic)
    'ViewMedia', 'UploadMedia',
    // System & Analytics (basic)
    'ViewDashboard'
  ],
  
  Viewer: [
    // Read-only access
    'ViewConversations', 'ViewMessages', 'ViewContacts', 'ViewTemplates', 'ViewMedia', 'ViewDashboard'
  ]
}

// Define base routes for each role to avoid circular dependencies
const AGENT_ROUTES = [
  '/app/dashboard',
  '/app/conversations',
  '/app/conversations/:id',
  '/app/contacts',
  '/app/templates',
  '/app/media/upload',
  '/app/profile',
  '/app/settings'
]

const SUPERVISOR_ROUTES = [
  '/app/dashboard',
  '/app/conversations',
  '/app/conversations/:id',
  '/app/conversations/assignment',
  '/app/conversations/monitor',
  '/app/team',
  '/app/analytics',
  '/app/performance',
  '/app/contacts',
  '/app/templates',
  '/app/flows',
  '/app/media',
  '/app/reports',
  '/app/profile',
  '/app/settings',
  ...AGENT_ROUTES
]

const ADMIN_ROUTES = [
  '/app/admin/*',
  '/app/admin/users',
  '/app/admin/system',
  '/app/admin/integrations',
  '/app/admin/webhooks',
  '/app/admin/audit',
  '/app/admin/billing',
  '/app/admin/organization',
  '/app/settings/advanced',
  ...SUPERVISOR_ROUTES
]

const VIEWER_ROUTES = [
  '/app/dashboard',
  '/app/analytics/view',
  '/app/reports/view',
  '/app/contacts/view',
  '/app/profile'
]

// Define routes accessible by each role
export const ROLE_ROUTES: Record<Role, string[]> = {
  Agent: AGENT_ROUTES,
  Supervisor: SUPERVISOR_ROUTES,
  Admin: ADMIN_ROUTES,
  Viewer: VIEWER_ROUTES
}

// Route-permission mapping
export const ROUTE_PERMISSIONS: Record<string, Permission[]> = {
  '/app/dashboard': ['ViewDashboard'],
  '/app/conversations': ['ViewConversations'],
  '/app/conversations/:id': ['ViewConversations', 'ViewMessages'],
  '/app/conversations/assignment': ['AssignConversations'],
  '/app/team': ['ViewUsers'],
  '/app/analytics': ['ViewReports'],
  '/app/contacts': ['ViewContacts'],
  '/app/templates': ['ViewTemplates'],
  '/app/flows': ['UpdateConversations'], // Flow management
  '/app/media': ['ViewMedia'],
  '/app/reports': ['ViewReports'],
  '/app/admin/users': ['ViewUsers', 'CreateUsers'],
  '/app/admin/system': ['ViewSystemSettings'],
  '/app/admin/integrations': ['ManageIntegrations'],
  '/app/admin/webhooks': ['ManageWebhooks'],
  '/app/admin/audit': ['ViewAuditLogs']
}

// Utility functions
export function hasPermission(userRole: Role, permission: Permission): boolean {
  const rolePermissions = ROLE_PERMISSIONS[userRole] || []
  return rolePermissions.includes(permission)
}

export function hasRole(userRole: Role, requiredRole: Role): boolean {
  const roleHierarchy: Record<Role, number> = {
    Viewer: 1,
    Agent: 2,
    Supervisor: 3,
    Admin: 4
  }
  
  return roleHierarchy[userRole] >= roleHierarchy[requiredRole]
}

export function canAccessRoute(userRole: Role, route: string): boolean {
  // Check if route matches any allowed routes for the role
  const allowedRoutes = ROLE_ROUTES[userRole] || []
  
  // Direct match
  if (allowedRoutes.includes(route)) {
    return true
  }
  
  // Pattern matching (e.g., /app/admin/* for admin routes)
  const routeMatches = allowedRoutes.some(allowedRoute => {
    if (allowedRoute.endsWith('/*')) {
      const baseRoute = allowedRoute.slice(0, -2)
      return route.startsWith(baseRoute)
    }
    
    // Parameter matching (e.g., /conversations/:id)
    if (allowedRoute.includes(':')) {
      const routeParts = route.split('/')
      const allowedParts = allowedRoute.split('/')
      
      if (routeParts.length !== allowedParts.length) {
        return false
      }
      
      return allowedParts.every((part, index) => {
        return part.startsWith(':') || part === routeParts[index]
      })
    }
    
    return false
  })
  
  if (routeMatches) {
    return true
  }
  
  // Check route permissions
  const requiredPermissions = ROUTE_PERMISSIONS[route] || []
  if (requiredPermissions.length === 0) {
    return true // No specific permissions required
  }
  
  return requiredPermissions.some(permission => hasPermission(userRole, permission))
}

export function getDefaultRoute(role: Role): string {
  switch (role) {
    case 'Agent':
      return '/app/conversations'
    case 'Supervisor':
      return '/app/dashboard'
    case 'Admin':
      return '/app/dashboard'
    case 'Viewer':
      return '/app/dashboard/readonly'
    default:
      return '/app/dashboard'
  }
}

export function getAllowedRoutes(role: Role): string[] {
  return ROLE_ROUTES[role] || []
}